providers/implementations/keymgmt/hybrid_kmgmt.c

```diff
@@ -0,0 +1,782 @@
+        }
+        data = OPENSSL_memdup(params->data, params->data_size);
+        if (data == NULL)
+            return 0;
+        pos = p - arg->params;
+        arg->d[pos][arg->alg_n]->data = data;
+        arg->d[pos][arg->alg_n]->data_size = params->data_size;
+    }
+    arg->alg_n++;
+    return 1;
+}
+
+/*
+ * Write out a length field.
+ * This is variable length.  The top two bits of the first byte determine the
+ * total number of trailing bytes.  The number is stored in big endian order.
+ */
+static unsigned char *write_length(unsigned char *p, size_t l)
+{
+    /* Determine the length 1 - 4 bytes */
+    const unsigned int n = 1 + !!(l >> 6) + !!(l >> 14) + !!(l >> 22);
+
+    switch (n) {
+    case 0:
+        *p++ = l & 077;
+        break;
+    case 1:
+        *p++ = (l >> 6 & 077) | 0100;
+        *p++ = l & 0xff;
+        break;
+    case 2:
+        *p++ = (l >> 14 & 077) | 0200;
+        *p++ = l >> 8 & 0xff;
+        *p++ = l & 0xff;
+        break;
+    case 3:
+        *p++ = (l >> 22 & 077) | 0300;
+        *p++ = l >> 16 & 0xff;
+        *p++ = l >> 8 & 0xff;
+        *p++ = l & 0xff;
+        break;
+    }
+    return p;
+}
+
+int ossl_hybrid_kmgmt_export(void *keydata, int selection,
+                             OSSL_CALLBACK *param_cb, void *cbarg)
+{
+    HYBRID_PKEY *key = (HYBRID_PKEY *)keydata;
+    HYBRID_EXPORT_CB_ARG arg;
+    unsigned int i, j;
+    int ret = 0;
+
+    if (!ossl_prov_is_running() || key == NULL)
+        return 0;
+
+    memset(&arg, 0, sizeof(arg));
+    for (i = 0; i < key->info->num_algs; i++)
+        if (!EVP_PKEY_export(key->keys[i], selection,
+                             &hybrid_export_cb, &arg))
+            goto err;
+
+    for (i = 0; i < arg.param_n; i++) {
+        size_t total = sizeof(size_t) * key->info->num_algs;
+        unsigned char *p;
+
+        for (j = 0; j < key->info->num_algs; j++)
+            if (arg.d[i][j]->data != NULL)
+                total += arg.d[i][j]->data_size;
+        p = OPENSSL_malloc(total);
+        if (p == NULL)
+            goto err;
+
+        arg.params[i].data = p;
+        for (j = 0; j < key->info->num_algs; j++) {
+            p = write_length(p, arg.d[i][j]->data_size);
+            if (arg.d[i][j]->data != NULL) {
+                memcpy(p, arg.d[i][j]->data, arg.d[i][j]->data_size);
+                p += arg.d[i][j]->data_size;
+            }
+        }
+        arg.params[i].data_size = p - (unsigned char *)arg.params[i].data;
+    }
+
+    arg.params[arg.param_n] = OSSL_PARAM_construct_end();
+    ret = param_cb(arg.params, cbarg);
+err:
+    /* Be conservative and clear everything in case it's private */
+    for (i = 0; i < arg.param_n; i++) {
+        OPENSSL_clear_free(arg.params[i].data, arg.params[i].data_size);
+        for (j = 0; j < arg.alg_n; j++)
+            OPENSSL_clear_free(arg.d[i][j]->data, arg.d[i][j]->data_size);
+    }
+    return ret;
+}
+
+const OSSL_PARAM *ossl_hybrid_ettable_common(const OSSL_PARAM *r) {
```

Comment:
Awkward name: `ossl_hybrid_getsettable_common()` would be perhaps more understandable?

Reply:
Yuk :grinning:  Every name I could think of was awkard.  At least `ettable` is pronounceable.